core/headerchain.go

```diff
@@ -175,71 +175,55 @@ func (hc *HeaderChain) WriteHeader(header *types.Header) (status WriteStatus, er
 	if err := headerBatch.Write(); err != nil {
 		log.Crit("Failed to write header into disk", "err", err)
 	}
-	// If the total difficulty is higher than our known, add it to the canonical chain
-	// Second clause in the if statement reduces the vulnerability to selfish mining.
-	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
-	reorg := externTd.Cmp(localTd) > 0
-	if !reorg && externTd.Cmp(localTd) == 0 {
-		if header.Number.Uint64() < head {
-			reorg = true
-		} else if header.Number.Uint64() == head {
-			reorg = mrand.Float64() < 0.5
+	// If the header can be added into canonical chain, adjust the
+	// header chain markers(canonical indexes and head header flag).
+	//
+	// Note all markers should be written atomically.
+
+	// Delete any canonical number assignments above the new head
+	markerBatch := hc.chainDb.NewBatch()
+	for i := number + 1; ; i++ {
+		hash := rawdb.ReadCanonicalHash(hc.chainDb, i)
+		if hash == (common.Hash{}) {
+			break
 		}
+		rawdb.DeleteCanonicalHash(markerBatch, i)
 	}
-	if reorg {
-		// If the header can be added into canonical chain, adjust the
-		// header chain markers(canonical indexes and head header flag).
-		//
-		// Note all markers should be written atomically.
 
-		// Delete any canonical number assignments above the new head
-		markerBatch := hc.chainDb.NewBatch()
-		for i := number + 1; ; i++ {
-			hash := rawdb.ReadCanonicalHash(hc.chainDb, i)
-			if hash == (common.Hash{}) {
+	// Overwrite any stale canonical number assignments
+	var (
+		headHash   = header.ParentHash
+		headNumber = header.Number.Uint64() - 1
+		headHeader = hc.GetHeader(headHash, headNumber)
+	)
+	for rawdb.ReadCanonicalHash(hc.chainDb, headNumber) != headHash {
```

Comment:
What is this for loop doing? Wouldn't we always expect to be adding on to HEAD? Why do we need to iterate backwards through the chain?

Reply:
Agree. Headerchain also refactored to avoid unnecessary checks